src/hotspot/share/prims/jni.cpp

```diff
@@ -2822,13 +2822,8 @@ JNI_ENTRY(void*, jni_GetPrimitiveArrayCritical(JNIEnv *env, jarray array, jboole
     *isCopy = JNI_FALSE;
   }
   oop a = lock_gc_or_pin_object(thread, array);
-  assert(a->is_array(), "just checking");
-  BasicType type;
-  if (a->is_objArray()) {
-    type = T_OBJECT;
-  } else {
-    type = TypeArrayKlass::cast(a->klass())->element_type();
-  }
+  guarantee(a->is_typeArray(), "Primitive array only");
```

Comment:
I think an assert here is more usual.  This is JNI - we don't make guarantees (pun slightly intentional).  If you want error reporting, use -Xcheck:jni; the appropriate check and reporting is already there.